grype/matcher/apk/matcher.go

```diff
@@ -135,3 +135,90 @@ func vulnerabilitiesByID(vulns []vulnerability.Vulnerability) map[string][]vulne
 
 	return results
 }
+
+func (m *Matcher) findApkPackage(store vulnerability.Provider, d *distro.Distro, p pkg.Package) ([]match.Match, error) {
+	// find Alpine SecDB matches for the given package name and version
+	secDbMatches, err := common.FindMatchesByPackageDistro(store, d, p, m.Type())
+	if err != nil {
+		return nil, err
+	}
+
+	cpeMatches, err := m.cpeMatchesWithoutSecDbFixes(store, d, p)
+	if err != nil {
+		return nil, err
+	}
+
+	var matches []match.Match
+
+	// keep all secdb matches, as this is an authoritative source
+	matches = append(matches, secDbMatches...)
+
+	// keep only unique CPE matches
+	matches = append(matches, deduplicateMatches(secDbMatches, cpeMatches)...)
+
+	return matches, nil
+}
+
+func (m *Matcher) matchBySourceIndirection(store vulnerability.Provider, d *distro.Distro, p pkg.Package) ([]match.Match, error) {
+	metadata, ok := p.Metadata.(pkg.ApkMetadata)
+	if !ok {
+		return nil, nil
+	}
+
+	// ignore packages without source indirection hints or where source name is identical to package name
+	if metadata.OriginPackage == "" || metadata.OriginPackage == p.Name {
+		return nil, nil
+	}
+
+	// build indirect package for matching against source package
+	indirectPackage, err := buildIndirectPackage(p, metadata)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build an indirect package for: %s", p.Name)
+	}
+
+	matches, err := m.findApkPackage(store, d, indirectPackage)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find vulnerabilities by apk source indirection: %w", err)
+	}
+
+	// we want to make certain that we are tracking the match based on the package from the SBOM (not the indirect package)
+	// however, we also want to keep the indirect package around for future reference
+	for idx := range matches {
+		matches[idx].Package = p
+
+		if matches[idx].Type == match.ExactDirectMatch {
+			matches[idx].Type = match.ExactIndirectMatch
+		}
+	}
+
+	return matches, nil
+}
+
+func buildIndirectPackage(p pkg.Package, metadata pkg.ApkMetadata) (pkg.Package, error) {
```

Comment:
Could we get the `metadata` value within this function's body (akin to what's being done in lines 163-171 in the current state), and then the function needs only the `p pkg.Package` parameter? My understanding is that `metadata` is only needed as a means of computing the indirect package...